UnitySDK/Assets/ML-Agents/Examples/SharedAssets/Scripts/RayPerception2D.cs

```diff
@@ -30,55 +31,63 @@ public class RayPerception2D : RayPerception
             float startOffset=0.0f, float endOffset=0.0f)
         {
-            m_PerceptionBuffer.Clear();
+
+            var perceptionSize = (detectableObjects.Length + 2) * rayAngles.Length;
+            if (m_PerceptionBuffer == null || m_PerceptionBuffer.Length != perceptionSize)
+            {
+                m_PerceptionBuffer = new float[perceptionSize];
+            }
+            Array.Clear(m_PerceptionBuffer, 0, m_PerceptionBuffer.Length);
+
             // For each ray sublist stores categorical information on detected object
             // along with object distance.
+            var bufferOffset = 0;
             foreach (var angle in rayAngles)
             {
-                m_EndPosition = transform.TransformDirection(
-                    PolarToCartesian(rayDistance, angle));
+                Vector2 rayDirection = transform.TransformDirection(PolarToCartesian(rayDistance, angle));
                 if (Application.isEditor)
                 {
                     Debug.DrawRay(transform.position,
-                        m_EndPosition, Color.black, 0.01f, true);
+                        rayDirection, Color.black, 0.01f, true);
                 }
 
-                var subList = new float[detectableObjects.Length + 2];
-                m_Hit = Physics2D.CircleCast(transform.position, 0.5f, m_EndPosition, rayDistance);
+                m_Hit = Physics2D.CircleCast(transform.position, 0.5f, rayDirection, rayDistance);
                 if (m_Hit)
                 {
                     for (var i = 0; i < detectableObjects.Length; i++)
                     {
                         if (m_Hit.collider.gameObject.CompareTag(detectableObjects[i]))
                         {
-                            subList[i] = 1;
-                            subList[detectableObjects.Length + 1] = m_Hit.distance / rayDistance;
+                            m_PerceptionBuffer[bufferOffset + i] = 1;
+                            m_PerceptionBuffer[bufferOffset + detectableObjects.Length + 1] = m_Hit.distance / rayDistance;
                             break;
                         }
                     }
                 }
                 else
                 {
-                    subList[detectableObjects.Length] = 1f;
+                    m_PerceptionBuffer[bufferOffset + detectableObjects.Length] = 1f;
                 }
 
-                m_PerceptionBuffer.AddRange(subList);
+                bufferOffset += detectableObjects.Length + 2;
             }
 
             return m_PerceptionBuffer;
         }
 
+
+
         /// <summary>
         /// Converts polar coordinate to cartesian coordinate.
         /// </summary>
         public static Vector2 PolarToCartesian(float radius, float angle)
         {
-            var x = radius * Mathf.Cos(DegreeToRadian(angle));
-            var y = radius * Mathf.Sin(DegreeToRadian(angle));
+            var x = radius * Mathf.Cos(RayPerceptionSensor.DegreeToRadian(angle));
```

Comment:
Same comment as in PolarToCartesian. Is there a better place to put these ?

Reply:
I removed DegreeToRadian since we can just do `Mathf.Deg2Rad * angleDegrees` instead

Reply:
Also made the PolarToCartesian methods private in their respective classes.

Reply:
This is folded into RayPerceptionSensor now.